test/react-web/client/ApolloProvider.test.tsx

```diff
@@ -103,4 +156,118 @@ describe('<ApolloProvider /> Component', () => {
+    const testClient = new ApolloClient();
+    testClient.store = store;
+
+    const initStoreMock = jest.fn();
+    testClient.initStore = initStoreMock;
+
+    const container = TestUtils.renderIntoDocument(
+      <Container client={testClient} />
+    ) as React.Component<any, any>;
+    expect(initStoreMock).toHaveBeenCalled();
+
+    initStoreMock.mockClear();
+    const newClient = new ApolloClient();
+    newClient.store = store;
+    newClient.initStore = initStoreMock;
+    container.setState({ client: newClient });
+
+    expect(initStoreMock).toHaveBeenCalled();
+  });
+
+  it('should not call clients init store when a store is passed', () => {
+    const testClient = new ApolloClient();
+
+    const initStoreMock = jest.fn();
+    testClient.initStore = initStoreMock;
+
+    const container = TestUtils.renderIntoDocument(
+      <Container client={testClient} store={store} />
+    ) as React.Component<any, any>;
+
+    expect(initStoreMock).not.toHaveBeenCalled();
+
+    initStoreMock.mockClear();
+    const newClient = new ApolloClient();
+    container.setState({ client: newClient });
+
+    expect(initStoreMock).not.toHaveBeenCalled();
+
+    const newStore = createStore(() => ({}));
+    container.setState({ store: store });
+    expect(initStoreMock).not.toHaveBeenCalled();
+  });
+
+  it('child component should be able to query new client and store when props change', () => {
+    const container = TestUtils.renderIntoDocument(
+      <Container client={client} store={store} />
+    ) as React.Component<any, any>;
+
+    const child = TestUtils.findRenderedComponentWithType(container, Child);
+    expect(child.context.client).toEqual(client);
+    expect(child.context.store).toEqual(store);
+
+    const newClient = new ApolloClient({});
+    const newStore = createStore(() => ({}));
+
+    container.setState({
+      client: newClient,
+      store: newStore,
+    });
+
+    expect(child.context.client).toEqual(newClient);
+    expect(child.context.store).toEqual(newStore);
+    expect(child.context.client).not.toEqual(client);
+    expect(child.context.store).not.toEqual(store);
+  });
+
+  it('should refetch against the new client when the client prop changes', () => {
+    const initialInterface = { query: jest.fn() };
+    const initialClient = new ApolloClient({
+      networkInterface: initialInterface,
+    });
+
+    const container = TestUtils.renderIntoDocument(
+      <ConnectedContainer client={initialClient} />
+    ) as React.Component<any, any>;
+
+    expect(initialInterface.query).toHaveBeenCalled();
+    initialInterface.query.mockClear();
+
+    const nextInterface = { query: jest.fn() };
+    const nextClient = new ApolloClient({
+      networkInterface: nextInterface,
+    });
+
+    container.setState({
+      client: nextClient,
+    });
+
+    // Both cases fail
+    expect(initialInterface.query).not.toHaveBeenCalled();
+    expect(nextInterface.query).toHaveBeenCalled();
```

Comment:
Hey @calebmer, revisited this problem after learning a bit more and running into a bug that fit the test case you described exactly in this PR. I'm wondering now if you foresaw this being an issue?

Any help/insight is appreciated.